theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,164 @@
+                'gpuarray_api.h', 'gpuarray/array.h', 'gpuarray/util.h']
+
+    def make_node(self, activations, labels, input_lengths):
+        """
+        Parameters
+        ----------
+        activations
+            Three-dimensional tensor, which has a shape of (t, m, p), where
+            t is the time index, m is the minibatch index, and p is the index
+            over the probabilities of each symbol in the alphabet. The memory
+            layout is assumed to be in C-order, which consists in the slowest
+            to the fastest changing dimension, from left to right. In this case,
+            p is the fastest changing dimension.
+        labels
+            A 1-D tensor of all the labels for the minibatch.
+        input_lengths
+            A 1-D tensor with the number of time steps for each sequence in
+            the minibatch.
+
+        """
+
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+
+        context = infer_context_name(activations, labels, input_lengths)
+        assert context == self.context_name
+
+        t_activations = as_gpuarray_variable(activations,
+                                             context_name=self.context_name)
+        # Ensure activations array is C-contiguous
+        t_activations = gpu_contiguous(t_activations)
+
+        # Labels and input lengths are always on the CPU
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.as_tensor_variable(input_lengths)
+
+        if t_activations.type.dtype != 'float32':
+            raise TypeError('Activations must use the float32 type!')
+
+        if t_labels.type.dtype != 'int32':
+            raise TypeError('Labels must use the int32 type!')
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('Label lengths must use the int32 type!')
+
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
+
+        out_params = [as_gpuarray_variable(self.costs(), context_name=self.context_name)]
+        if self.gradients is not None:
+            out_params.append(as_gpuarray_variable(self.gradients(),
+                                                   context_name=self.context_name))
+
+        return theano.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                            outputs=out_params)
+
+    def grad(self, inputs, output_grads):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        z = output_grads[0]
+        grad_shuffle = GpuDimShuffle(input_broadcastable=(False, False, False),
+                                     new_order=(1, 0, 2))(self.gradients())
```

Comment:
As info, I think your errors occur because `self.gradients()` here is never included in a graph: in `__init__()`, `self.gradients` is defined as a Theano type. So, call to `self.gradients()` creates a new variable in `make_node()`, and the same call here creates another nex variable that is not used anywhere else.

But as we completely change implementation, this error should be avoided.